jupyterhub/handlers/pages.py

```diff
@@ -106,22 +106,27 @@ async def _render_form(self, for_user, spawner_options_form, message=''):
         )
 
     @web.authenticated
-    async def get(self, for_user=None, server_name=''):
+    def get(self, user_name=None, server_name=''):
         """GET renders form for spawning with user-specified options
 
         or triggers spawn via redirect if there is no form.
         """
+        # two-stage to get the right signature for @require_scopes filter on user_name
+        if user_name is None:
+            user_name = self.current_user.name
+        if server_name is None:
+            server_name = ""
+        return self._get(user_name=user_name, server_name=server_name)
 
```

Comment:
Is the process of returning a coroutine here what makrs you remove the async keyword from this get handler - in practice "making it async" anyhow by returning a coroutine?

Reply:
Yes, a coroutine object is an object like anything else that can be returned from any function, async or not. I think this was clearer in the `@coroutine` decorator days, where it was more apparent that it really must be a regular Python function returning a regular Python object. The `async def` syntax can feel like something special is happening, when it's really not. The key, I think: calling a **coroutine function** returns a **coroutine object** which is **awaitable** (other awaitables include Tasks, Futures, any object implementing `__await__`, etc.). Any function can return any kind of object, so if you wrap a function and return its original return object, calling the wrapper will have the same behavior as the original wrapped function.

 So if you 'wrap' a coroutine, the wrapper does not itself need to be async unless it, too, has something to await. I could have made this _logically_ equivalent with `async def` if it ended with `return await self._get(...)`, but that would add an unnecessary callback layer to the event loop.

Reply:
Thanks for helping me learn about the details Min!